### ISSUE TITLE
YamlFrontMatterBlock rendered as markdown does not preserve nested lists

### ISSUE THREAD

**Issue by olafurpg** [vsch/flexmark-java#293]
I am using `Parser` and `Formatter` to transform markdown documents and produce markdown documents. The problem is that YAML frontmatter when using `YamlFrontmatterExtension` with nested lists are not preserved when rendering back into markdown.

* [x] `Parser`
* [x] `Formatter`
* [x] `YamlFrontmatterExtension`

A frontmatter block like this
```md
---
linters:
  - name: linter
    rules:
      - name: UseIsNanNotNanComparison
        url:  https://github.com/HairyFotr/linter/blob/master/src/test/scala/LinterPluginTest.scala#L1930
---
```

renders into the following output
```md
---
linters:
- name: linter
rules:
- name: UseIsNanNotNanComparison
url:  https://github.com/HairyFotr/linter/blob/master/src/test/scala/LinterPluginTest.scala#L1930
---
```

I expected the rendered output to be the same as the input.

Minimal reproduction code example in Scala.

```scala
    val settings = new MutableDataSet()
      .set(Parser.EXTENSIONS, List(YamlFrontMatterExtension.create()).asJava)
    val parser = Parser.builder(settings).build()
    val formatter = Formatter.builder(settings).build()
    val markdown =
      """|---
         |linters:
         |  - name: linter
         |    rules:
         |      - name: UseIsNanNotNanComparison
         |        url:  https://github.com/HairyFotr/linter/blob/master/src/test/scala/LinterPluginTest.scala#L1930
         |---
      """.stripMargin
    val parsedMarkdown = parser.parse(markdown)
    val renderedMarkdown = formatter.render(parsedMarkdown)
    assertNoDiff(renderedMarkdown, markdown)
```

A workaround is to unlink the frontmatter block and prepend it's chars to rendered output.

```scala
    val appendable = new java.lang.StringBuilder()
    document.getChildren.asScala.headOption match {
      case Some(f: YamlFrontMatterBlock) =>
        appendable.append(f.getChars)
        f.unlink()
      case _ =>
    }
    formatter.render(document, appendable)
    appendable.toString
```

**Comment by vsch** [comment]
@olafurpg, I missed setting pre-formatted for yaml block so whitespace is preserved.

I also added first level values as child nodes of yaml node so proper offsets in file can be determined.

* Fix: #293, YamlFrontMatterBlock rendered as markdown does not preserve nested lists
  * Add: `YamlFrontMatterValue` node containing yaml value(s) inserted as children of
    `YamlFrontMatterNode`
  * Fix: change stored key as `BasedSequence` instead of string, can be retrieved as `YamlFrontMatterNode.getKeySequence()`
  * Add: code to return `List<String>` from child nodes of `YamlFrontMatterNode`
  * Fix: resolve offsets in `YamlFrontMatterNode` and `YamlFrontMatterValue` nodes.

The changes do not properly parse list values, this is a limitation of the original yaml parser implementation and without rewriting it for full Yaml syntax cannot be fixed. I would not recommend using the block key/values parsed for anything but simple values.

**Comment by olafurpg** [comment]
I am OK with the limitation that `YamlFrontMatterNode` doesn't support all of YAML (it's a huge spec!). This issue is primarily about preserving the original formatting of the frontmatter block when the `YamlFrontMatterNode` is unchanged. In my particular use-case, I only transform code fences.

**Comment by vsch** [comment]
@olafurpg, the issue is fixed in 0.40.4.

**Comment by olafurpg** [comment]
Thank you @vsch I am unable to reproduce the bug in 0.40.4, great job!
